addr_manager_ds.go

```diff
@@ -0,0 +1,316 @@
+				log.Error(err)
+				continue
+			}
+			keys = append(keys, key)
+
+			if ttl <= 0 {
+				batch.Delete(key)
+				continue
+			}
+			has, err := mgr.ds.Has(key)
+			if err != nil || !has {
+				mgr.subsManager.BroadcastAddr(p, addr)
+			}
+
+			// Allows us to support AddAddr and SetAddr in one function
+			if !has || !add {
+				if err := batch.Put(key, addr.Bytes()); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+		if err := batch.Commit(); err != nil {
+			log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+			continue
+		}
+		mgr.ttlManager.setTTLs(keys, ttl)
+		return
+	}
+	log.Errorf("failed to avoid write conflict for peer %s after %d retries\n", p.Pretty(), dsWriteRetries)
+}
+
+// UpdateAddrs will update any addresses for a given peer and TTL combination to
+// have a new TTL.
+func (mgr *DatastoreAddrManager) UpdateAddrs(p peer.ID, oldTTL time.Duration, newTTL time.Duration) {
+	prefix := ds.NewKey(p.Pretty())
+	mgr.ttlManager.updateTTLs(prefix, oldTTL, newTTL)
+}
+
+// Addrs Returns all of the non-expired addresses for a given peer.
+func (mgr *DatastoreAddrManager) Addrs(p peer.ID) []ma.Multiaddr {
+	prefix := ds.NewKey(p.Pretty())
+	q := query.Query{Prefix: prefix.String()}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []ma.Multiaddr{}
+	}
+
+	var addrs []ma.Multiaddr
+	for result := range results.Next() {
+		addrbytes := result.Value.([]byte)
+		addr, err := ma.NewMultiaddrBytes(addrbytes)
+		if err != nil {
+			continue
+		}
+		addrs = append(addrs, addr)
+	}
+
+	return addrs
+}
+
+// Peers returns all of the peer IDs for which the AddrBook has addresses.
+func (mgr *DatastoreAddrManager) Peers() []peer.ID {
+	q := query.Query{}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []peer.ID{}
+	}
+
+	idset := make(map[peer.ID]struct{})
+	for result := range results.Next() {
+		key := ds.RawKey(result.Key)
+		id, err := peerIDFromKey(key)
+		if err != nil {
+			continue
+		}
+		idset[id] = struct{}{}
+	}
+
+	ids := make([]peer.ID, 0, len(idset))
+	for id := range idset {
+		ids = append(ids, id)
+	}
+	return ids
+}
+
+// AddrStream returns a channel on which all new addresses discovered for a
+// given peer ID will be published.
+func (mgr *DatastoreAddrManager) AddrStream(ctx context.Context, p peer.ID) <-chan ma.Multiaddr {
+	initial := mgr.Addrs(p)
+	return mgr.subsManager.AddrStream(ctx, p, initial)
+}
+
+// ClearAddrs will delete all known addresses for a peer ID.
+func (mgr *DatastoreAddrManager) ClearAddrs(p peer.ID) {
+	prefix := ds.NewKey(p.Pretty())
+	for i := 0; i < dsWriteRetries; i++ {
+		q := query.Query{Prefix: prefix.String()}
+		results, err := mgr.ds.Query(q)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+		batch, err := mgr.ds.Batch()
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		for result := range results.Next() {
+			batch.Delete(ds.NewKey(result.Key))
+		}
+		if err = batch.Commit(); err != nil {
```

Comment:
style preference: whenever using this error handling pattern, use `if err := thing();` instead of `if err = thing();`